armi/utils/gridEditor.py

```diff
@@ -0,0 +1,1783 @@
+        )
+
+        self.parent.setMode(mode)
+
+    def editorClicked(self):
+        self.pathControl.maybeIncrement()
+
+    def getSelectedAssem(self) -> Optional[Union[AssemblyBlueprint, Tuple[int, int]]]:
+        """
+        Return the currently-selected assembly design or fuel path indices
+        """
+        if self.activeAssemID in self.assemDesignsById:
+            # We have an assembly design activated. return it
+            return self.assemDesignsById[self.activeAssemID]
+        elif self.activeAssemID in self.pathControl.getActivateButtons():
+            # we are in path selection mode, return stuff from the pathControl
+            return self.pathControl.getIndices(clear=False)
+        else:
+            return None
+
+    def getAssemToSet(self) -> Optional[Union[AssemblyBlueprint, Tuple[int, int]]]:
+        """
+        Return the assembly design of fuel path tuple that a client should set.
+
+        This differs from ``getSelectedAssem`` in that it can incorporate more logic to
+        enforce certain rules, such as performing increments, masking things off based
+        on other state etc., whereas ``getSelectedAssem`` should be more dumb and just
+        return the state of the controls themselves.
+        """
+        if self.activeAssemID in self.assemDesignsById:
+            # We have an assembly design activated. return it
+            return self.assemDesignsById[self.activeAssemID]
+        elif self.activeAssemID in self.pathControl.getActivateButtons():
+            # we are in path selection mode, return stuff from the pathControl
+            return self.pathControl.getIndices(clear=True)
+        else:
+            return None
+
+    def setActiveAssem(self, assemDesign: Optional[Union[AssemblyBlueprint, tuple]]):
+        """
+        Override the selected assembly design from above.
+        """
+        specifier = None
+        if isinstance(assemDesign, AssemblyBlueprint):
+            specifier = assemDesign.specifier
+            self._setActiveAssemID(self.buttonIdBySpecifier[specifier])
+        elif isinstance(assemDesign, tuple):
+            self._setActiveAssemID(self.pathControl.activateButton.GetId())
+        elif assemDesign is None:
+            self._setActiveAssemID(None)
+
+
+class GridGui(wx.ScrolledWindow):
+    """
+    Visual editor for grid blueprints.
+
+    This is the actual viewer that displays the grid and grid blueprints contents, and
+    responds to mouse events. Under the hood, it uses a wx.PseudoDC to handle the
+    drawing, which provides the following benefits over a regular DC:
+     - Drawn objects can be associated with an ID, allowing parts of the drawing to be
+       modified or cleared without having to re-draw everything.
+     - The IDs associated with the objects can be used to distinguish what was clicked
+       on in a mouse event (though the support for this isn't super great, so we do have
+       to do some of our own object disambiguation).
```

Comment:
No, from what i can tell, the pdc stuff is either buggy, or poorly described. It's been a while since i originally wrote this stuff, but if memory serves, the hit boxes of objects are all rectangles, so for something like a hex, it's possible click on two at once. when that happens you get multiple ids (i think you're right, in the order that they are created), but still have to do something to figure out which one is right. I spent quite a bit of time playing with this, but its possible that i missed something